src/Interpreters/Cache/FileCache.cpp

```diff
@@ -170,6 +188,35 @@ void FileCache::assertInitialized() const
 }
 
 void FileCache::initialize()
+{
+    // Prevent initialize() from running twice. This may be caused by two cache disks being created with the same path (see integration/test_filesystem_cache).
+    callOnce(initialize_called, [&] {
+        bool need_to_load_metadata = fs::exists(getBasePath());
+        try
+        {
+            if (!need_to_load_metadata)
+              fs::create_directories(getBasePath());
+            status_file = make_unique<StatusFile>(fs::path(getBasePath()) / "status", StatusFile::write_full_info);
+        }
+        catch (...)
+        {
+            init_exception = std::current_exception();
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+            throw;
+        }
+
+        if (load_metadata_asynchronously)
+        {
+            load_metadata_main_thread = ThreadFromGlobalPool([this, need_to_load_metadata] { initializeImpl(need_to_load_metadata); });
+        }
+        else
+        {
+            initializeImpl(need_to_load_metadata);
```

Comment:
Exceptions from `initializeImpl` should also be caught and set to `init_exception`, both here and if load_metadata_asynchronously.

Reply:
Exceptions from the call to `loadMetadata()` in `initializeImpl()` are handled that way inside `initializeImpl()`. The call to `metadata.startup()` is not handled that way as it was the previous behavior, but I guess it makes sense to include it in the `try` block too, I've added it

Reply:
Indeed. But now there can be a data race on `init_exception`. Please add a mutex.

Reply:
What's the behavior that you'd expect in `isInitialized()`? I'm thinking it's good to avoid a data race on `init_exception`, but it should not block if  `init_mutex` is taken. WDYT about this?:
```cpp
        std::unique_lock lock(init_mutex, std::try_to_lock);
        if (lock)
        {
            /// Exception during initialization,
            /// make user aware that the cache is broken.
            if (init_exception && throw_on_cache_initialization_error)
            {
                std::rethrow_exception(init_exception);
            }
        }
        return is_initialized;
```

Reply:
something like this
```
        if (is_initialized)
            return true;
        if (!throw_on_cache_initialization_error)
            return false;
        std::lock_guard(init_mutex);
        if (init_exception)
            std::rethrow_exception(init_exception);
        return is_initiailized;
```

Reply:
`assertInitialized` anyway takes the lock as well: https://github.com/ClickHouse/ClickHouse/blob/7e9e03b9c67d14878fb1bbf8b8eb5aa047a2327a/src/Interpreters/Cache/FileCache.cpp#L157-L163

Reply:
I think we're overcomplicating `isInitialized()` with logic that is hardly expected in a getter. In the latest commit I've made a different proposal:
- `isInitialized()` simply returns `is_initialized`,
- added a method `initException()` that is a getter for `init_exception`.

Rationale for this proposal is:
- `isInitialized()` is simplified
- `throw_on_cache_initialization_error` setting is not necessary
- Code that only cares about whether the cache is initialized (`ReadBufferFromRemoteFSGather`) can use this method without handling exceptions
- Code that wants to handle the exception, such as `InterpreterSystemQuery`, can get the exception to do so

The only thing that it doesn't have is the behavior that you requested in https://github.com/ClickHouse/ClickHouse/pull/65736#discussion_r1657314071, could we reconsider this? I find it inconsistent that if queries work by reading from cloud storage directly while the cache is in process of initialization, they start failing once an error is found. And we are already logging an error in `initializeImpl()`.

If you still think it is necessary for queries to throw if the cache has failed to initialize, we can add code similar to the one in `InterpreterSystemQuery` to check and throw `init_exception`, possibly in a helper method like the one you proposed (that doesn't block on `init_mutex`, as that would render the `load_metadata_asynchronously` setting useless, queries would have to wait for the cache in any case)

Reply:
I've removed `initException()` since it could cause a data race if two threads want to throw the exception simultaneously

Reply:
> The only thing that it doesn't have is the behavior that you requested in https://github.com/ClickHouse/ClickHouse/pull/65736#discussion_r1657314071, could we reconsider this? I find it inconsistent that if queries work by reading from cloud storage directly while the cache is in process of initialization, they start failing once an error is found. And we are already logging an error in initializeImpl().

If asynchronous loading if metadata is turned on - yes, it is ok not to rethrow the exception ever, so the user will always read from object storage. But in case it is disabled, which is the default, then we want to have old behaviour, that if cache failed to initialize on server startup, then it is visible to the user by default.

Reply:
> If asynchronous loading if metadata is turned on - yes, it is ok not to rethrow the exception ever, so the user will always read from object storage.

In this case no need to worry about data race on init_exception, no mutex needed.